{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e6f9064",
   "metadata": {},
   "outputs": [],
   "source": [
    " '''\n",
    "Code to extract barcodes from pKH106 integrated into the yeast genome. Works on single read data (ie SR100 NovaSeqX runs).\n",
    "\n",
    "The code assumes that you've sequenced the data with either:\n",
    "\n",
    "Primer pair 1:\n",
    "\n",
    "NGS R1 604 bcd xN F1    CACTCTTTCCCTACACGACGCTCTTCCGATCT NNNNNNNN GGATCCGATCATGCTT\n",
    "NGS R1 604 bcd 8N R1-2  TGACTGGAGTTCAGACGTGTGCTCTTCCGATCT NNNNNNNN CAGAAAACGTCATGGAG\n",
    "\n",
    "OR\n",
    "\n",
    "Primer pair 2:\n",
    "\n",
    "NGS R1 604 bcd 8N F2    CACTCTTTCCCTACACGACGCTCTTCCGATCT NNNNNNNN GGATAAAATGTGATAACTAATCAGC\n",
    "NGS R1 604 bcd 6N R2-2  TGACTGGAGTTCAGACGTGTGCTCTTCCGATCT NNNNNNNN AAACAATCAATGCCAGAG\n",
    "\n",
    "\n",
    "You need to specify which one was used.\n",
    "\n",
    "'''\n",
    "\n",
    "#### Data files\n",
    "\n",
    "\n",
    "dataDir = \"/file/directory/\"\n",
    "dataPrefix = \"da-ta-pre-fix\"\n",
    "fastqList = ['fast','q','file','identifiers']\n",
    "\n",
    "             \n",
    "             \n",
    "dataPostfix = \"da-ta-post-fix.gz\"\n",
    "\n",
    "F2primers = True # Is the data with the F2 + R2-2 primers? If False, assumed to be F1 + R1-2\n",
    "\n",
    "\n",
    "#### Code\n",
    "\n",
    "import regex\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "\n",
    "import scipy\n",
    "import numpy as np\n",
    "from scipy.sparse import dok_array, save_npz\n",
    "\n",
    "\n",
    "def encodeATGC(bcdStr):\n",
    "    return int(bcdStr.replace(\"A\",\"0\").replace(\"T\",\"1\").replace(\"G\",\"2\").replace(\"C\",\"3\"), 4)\n",
    "\n",
    "\n",
    "def encodeATGC_revComp(bcdStr):\n",
    "    bcdStr = bcdStr[::-1]\n",
    "    return int(bcdStr.replace(\"A\",\"1\").replace(\"T\",\"0\").replace(\"G\",\"3\").replace(\"C\",\"2\"), 4)\n",
    "\n",
    "def decodeATGC(bcdInt, bcdLength = 12):\n",
    "    bcdString = np.base_repr(bcdInt, base = 4)\n",
    "    bcdString = (\"0\"*(bcdLength-len(bcdString)) ) + bcdString\n",
    "    return bcdString.replace(\"0\",\"A\").replace(\"1\",\"T\").replace(\"2\",\"G\").replace(\"3\",\"C\")\n",
    "\n",
    "\n",
    "\n",
    "bcdlength = 12\n",
    "umilength_1 = 8\n",
    "\n",
    "##### Specifying Regular expressions\n",
    "\n",
    "# F1 + R1-1 Primer Regex\n",
    "UMI = \"([A|T|G|C]{\" + str(umilength_1) + \"})\"\n",
    "primer = \"(GGATCCGATCATGCTT){e<=3}\"\n",
    "barcode = \"([A|T|G|C]{4}(TT){e<=1}[A|T|G|C]{4}(TT){e<=1}[A|T|G|C]{4})\"\n",
    "threeprime = \"(GGTACCGCTGATTAGT){e<=3}\"\n",
    "\n",
    "read1regex_F1 = regex.compile(UMI + primer + barcode + threeprime)\n",
    "\n",
    "\n",
    "# F2 + R2-2 Primer regex\n",
    "UMI = \"([A|T|G|C]{\" + str(umilength_1) + \"})\"\n",
    "primer = \"(GGATAAAATGTGATAACTAATCAGCGGTACC){e<=4}\"\n",
    "barcode = \"([A|T|G|C]{4}(AA){e<=1}[A|T|G|C]{4}(AA){e<=1}[A|T|G|C]{4})\"\n",
    "threeprime = \"(AAGCATGATCGGATCC){e<=3}\"\n",
    "\n",
    "read1regex_F2 = regex.compile(UMI + primer + barcode + threeprime)\n",
    "\n",
    "\n",
    "if F2primers:\n",
    "    encodeBcd = encodeATGC_revComp\n",
    "    read1regex = read1regex_F2\n",
    "else:\n",
    "    encodeBcd = encodeATGC\n",
    "    read1regex = read1regex_F1\n",
    "\n",
    "\n",
    "for fastq in fastqList:\n",
    "    print(\"Starting \" + fastq)\n",
    "    fname = dataPrefix + fastq + dataPostfix\n",
    "\n",
    "    bcdUMICounts = dok_array((4**bcdlength, 4**(umilength_1)), dtype=np.uint64)\n",
    "    bcdCounts = dok_array((4**bcdlength, 1), dtype=np.uint64)\n",
    "\n",
    "    allseq = 0\n",
    "    bcdseq = 0\n",
    "\n",
    "    with gzip.open(dataDir + fname, \"rt\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fastq\"):\n",
    "            allseq = allseq + 1\n",
    "            m = read1regex.match(str(record.seq))\n",
    "            if m:\n",
    "                bcd = m.groups()[2][0:4] + m.groups()[2][6:10] + m.groups()[2][12:16] \n",
    "                if \"N\" not in bcd:\n",
    "                    bcdseq = bcdseq + 1\n",
    "                    umi = m.groups()[0]\n",
    "                \n",
    "                    bcdUMICounts[encodeBcd(bcd), encodeATGC(umi)] = bcdUMICounts[encodeBcd(bcd), encodeATGC(umi)] + 1\n",
    "                    bcdCounts[encodeBcd(bcd), 0] = bcdCounts[encodeBcd(bcd), 0] + 1\n",
    "\n",
    "                \n",
    "    print(fastq)            \n",
    "    print(\"Total reads: \" + str(allseq))\n",
    "    print(\"Reads passing quality control: \" + str(bcdseq))\n",
    "    print(\"i.e. \" + str(100*bcdseq/allseq) + \"%\")\n",
    "\n",
    "    save_npz(fastq.split(\"S\")[0] + \"bcdUMI.npz\",bcdUMICounts.tocsr())\n",
    "    save_npz(fastq.split(\"S\")[0] + \"bcd.npz\",bcdCounts.tocsr())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
